activitysim/core/configuration/filesystem.py

```diff
@@ -0,0 +1,828 @@
+
+                # if inheriting, continue and backfill settings from the next existing settings file configs_dir_list
+
+                inherit_settings = s.get("inherit_settings")
+                if isinstance(inherit_settings, str):
+                    inherit_file_name = inherit_settings
+                    assert (
+                        os.path.join(dir, inherit_file_name) not in source_file_paths
+                    ), f"circular inheritance of {inherit_file_name}: {source_file_paths}: "
+                    # make a recursive call to switch inheritance chain to specified file
+
+                    logger.debug(
+                        "inheriting additional settings for %s from %s"
+                        % (file_name, inherit_file_name)
+                    )
+                    s, source_file_paths = self.read_settings_file(
+                        inherit_file_name,
+                        mandatory=True,
+                        include_stack=source_file_paths,
+                        configs_dir_list=configs_dir_list,
+                    )
+
+                    # backfill with the inherited file
+                    settings = backfill_settings(settings, s)
+                    break  # break the current inheritance chain (not as bad luck as breaking a chain-letter chain?...)
+
+        if len(source_file_paths) > 0:
+            settings["source_file_paths"] = source_file_paths
+
+        if mandatory and not settings:
+            raise SettingsFileNotFoundError(file_name, configs_dir_list)
+
+        # Adds proto_ suffix for disaggregate accessibilities
+        if args.SUFFIX is not None and args.ROOTS:
+            settings = suffix_tables_in_settings(settings, args.SUFFIX, args.ROOTS)
+
+        # we don't want to actually have inherit_settings as a settings
+        settings.pop("inherit_settings", None)
+
+        if validator_class is not None:
+            settings = validator_class.parse_obj(settings)
+
+        if include_stack:
+            # if we were called recursively, return an updated list of source_file_paths
+            return settings, source_file_paths
+
+        else:
+            return settings
+
+    def read_model_settings(
+        self,
+        file_name,
+        mandatory=False,
+    ):
+        # in the legacy implementation, this function has a default mandatory=False
+        return self.read_settings_file(file_name, mandatory=mandatory)
+
+    def read_model_spec(self, file_name: str):
+        from activitysim.core import simulate
+
+        return simulate.read_model_spec(self, file_name)
+
+    def read_model_coefficients(self, model_settings=None, file_name=None):
+        from activitysim.core import simulate
+
+        return simulate.read_model_coefficients(
+            self, model_settings=model_settings, file_name=file_name
+        )
+
+    def get_segment_coefficients(self, model_settings, segment_name):
+        from activitysim.core import simulate
```

Comment:
Moving all the imports to the top would be great!  And it's definitely on the to-do list. Unfortunately this code has gotten wound around itself in my effort to make these changes substantially backward compatible and keep everything working through the transition, such that putting the imports at the top level triggers a circular import that crashes ActivitySim.

Reply:
Ok, sounds like an issue should be created for this since it's not going to be addressed here? It would be good if the issue could provide a little bit more detail about the problem area(s) and temporary fixes for developers to try if they run into the issue (switching the import order of some packages, etc.)